src/Symfony/Component/DependencyInjection/Compiler/AttributeAutoconfigurationPass.php

```diff
@@ -0,0 +1,53 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\DependencyInjection\Compiler;
+
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+
+/**
+ * @author Alexander M. Turek <me@derrabus.de>
+ */
+final class AttributeAutoconfigurationPass implements CompilerPassInterface
+{
+    public function process(ContainerBuilder $container): void
+    {
+        if (80000 > \PHP_VERSION_ID) {
+            return;
+        }
+
+        $autoconfiguredAttributes = $container->getAutoconfiguredAttributes();
+
+        foreach ($container->getDefinitions() as $definition) {
+            if (!$definition->isAutoconfigured()) {
+                continue;
+            }
+
+            if (!$class = $container->getParameterBag()->resolveValue($definition->getClass())) {
+                continue;
+            }
+
+            try {
+                $reflector = new \ReflectionClass($class);
+            } catch (\ReflectionException $e) {
+                continue;
+            }
+
+            foreach ($reflector->getAttributes() as $attribute) {
+                if (!($configurator = $autoconfiguredAttributes[$attribute->getName()] ?? null)) {
+                    continue;
+                }
+
+                $configurator($definition, $attribute->newInstance());
```

Comment:
This is going to be my most important comment:

Configuring the definition directly is a no-go: we must give higher precedence to explicit configuration.

Fortunately, we already built such a system: instance-of-conditionals fill exactly this need.

To me, autoconfigurators should only be able to mutate these conditionals.

In order to implement this restricted capability, my proposal would be to rely on #39804:
autoconfigurators could return an instance of the `Autoconfigure` attribute class, which is the exact data structure allowed by instance-of-conditionals.

Then the pass would map this returned attribute object to the definition by using `$definition->setInstanceOfConditionals()`.

Reply:
You are right, passing in the whole definition is probably a bad idea.

We could instead pass an empty `ChildDefinition` or make the closure return one. That would be closer to how `ResolveInstanceofConditionalsPass` already works.

Reply:
We could try with a `ChildDefinition` yes. The drawback would be to allow configuring more than what instance-of-conditionals handle, while `Autoconfigure` exactly fits what is acceptable.

Reply:
BTW, I think we should pass the reflector instance to the auconfigurator: we can anticipate that having the `ReflectionParameter` will be required when we'll handle attributes on arguments. But I'm sure it could be useful to provide the ReflectionClass also, eg to read related attributes, or anything really.

Reply:
(see https://github.com/symfony/symfony/issues/39988)

Reply:
So, what we need is a data object that holds exactly the cases allowed by instance-of-conditionals and `Autoconfigure` of #39804 is such a data object, correct? What I'm unclear about is why we need `Autoconfigure` as a userland attribute. It can be just a data object that is returned by this closure and passed in the yaml file loader in the `AttributeAutoconfigurationPass`, can't it?

My whole idea when I sketched this attribute example in https://github.com/symfony/symfony/pull/39776#issuecomment-762354357 is to enforce no coupling between attributes and the DI component. Having an `Autoconfigure` tag does defeat that purpose and I'm unsure why (in my mind, it opens the door for using the attribute for cases we don't want - the cases where autoconfiguration based on interface should be used e.g. the `ResetInterface`).

Reply:
You're right about the data object: we need one, and it can be either a `ChildDefinition` or an instance of `Autoconfigure` from #39804, or something else.

About the rest of your comment, please comment on #39804 so that we can have all arguments about it in one place. Please also take the time to read the description and the implementation again as they might have changed since you last reviewed it.